### DISCUSSION TITLE
install mosparo under subpath

### DISCUSSION THREAD

**Discussion by davd-edia** [mosparo/mosparo#216] [Q&A] [answered]
Hey there. Wondering if it's possible to have mosparo served under a subpath, like domain.com/mosparo?
I'm planning on using a reverse proxy.

Thanks.

**Comment by zepich** [comment]
Hi @davd-edia 

Thank you very much for your question.

That should be possible, yes. If you're using Apache, we've already added the .htaccess file in the root directory, which should rewrite the requests correctly. If you're using nginx or a different webserver/reverse proxy, you would have to add the rewrite rules as we have in the `.htaccess` file (https://github.com/mosparo/mosparo/blob/master/.htaccess).

We recommend using a subdomain. The problem with the subdirectory installation is that we have a lot of files (like configuration files) that should not be exposed to the public, which could happen if the webserver/proxy is misconfigured. The main configuration file can't be exposed since it is a `.php` file (`config/env.mosparo.php`). For example, if you accidentally expose any configuration files, an attacker can access them and determine which version of mosparo you use.

Please let me know if you experience any issues or need help.

Kind regards,

zepich

**Reply by davd-edia** [reply]
Hey @zepich ! Thanks for the tip. Actually I'm using the docker composer setup... I think it uses nginx builtin, and then I have my reverse proxy.

**Reply by davd-edia** [reply]
so I've replaced the nginx conf and it's replacing all url's and prepending the subpath. But the wizards are all based on 302 redirects with Location headers pointing to root... so these don't work...

**Reply by zepich** [reply]
Hi @davd-edia 

Sorry for the delay.

If you're using the Docker image, you don't need the redirects since they are in the nginx config in the Docker image. So, you simply have to configure the proxy without the redirections (I guess, I haven't tested it).

The redirects I mentioned occur if your virtual host exposes all files (not just those in the `public` directory).

If it's still not working, please share your nginx configuration so I can have a look.

Thank you very much.

Kind regards,

zepich

**Reply by davd-edia** [reply]
Hey @zepich. What I mean by redirects are builtin. For example when you save a project you're redirected to the following page. Many times the button does not open a url, the browser is redirected by a 302 and location header. So I don't know anyway of avoiding this.

**Reply by zepich** [reply]
Hi @davd-edia 

Ah, sorry, I misunderstood you.

Can you maybe share the reverse proxy (nginx?) configuration so I can test it on my server?

Kind regards,

zepich

**Reply by davd-edia** [reply]
ok, so my setup has IIS as a reverse proxy pointing webserver/mosparo to http://mosparo_ip.

My lastest mosparo nginx conf is this:
```
server {
    listen 80 default_server;
    listen [::]:80 default_server;

    server_name _;
    root /mosparo/public;

    index index.php;

    error_log /var/log/nginx/project_error.log;
    access_log /var/log/nginx/project_access.log;

    sub_filter_types text/html text/css application/javascript;
    sub_filter_once off;
    sub_filter '="/build' '="/mosparo/build';
    sub_filter 'url(/build' 'url(/mosparo/build';
    sub_filter '="/"' '="/mosparo/"';
    sub_filter '="/setup' '="/mosparo/setup';
    sub_filter '="/login"' '="/mosparo/login"';
    sub_filter '="/password"' '="/mosparo/password"';
    sub_filter '="/administration' '="/mosparo/administration';
    sub_filter '="/account' '="/mosparo/account';
    sub_filter '="/logout' '="/mosparo/logout';
    sub_filter '="/project' '="/mosparo/project';

    proxy_redirect /project /mosparo/project;

    location = /favicon.ico {
        log_not_found off;
        access_log off;
    }

    location = /robots.txt {
        allow all;
        log_not_found off;
        access_log off;
    }

    location / {
        try_files $uri $uri/ /index.php?$args;
    }

    location ~ ^/resources/(.*)/(.*)\.css$ {
        try_files $uri /index.php?$args;
        log_not_found off;
        expires max;
    }

    location ~ ^/index\.php(/|$) {
        include fastcgi_params;
        fastcgi_intercept_errors on;
        fastcgi_pass 127.0.0.1:9000;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
    }

    location ~* \.(js|png|jpg|jpeg|gif|ico)$ {
        expires max;
        log_not_found off;
    }
}
```

As you can see I do a series of replaces of root references to /mosparo.

**Reply by zepich** [reply]
Hi @davd-edia 

Thank you very much for your configuration. I've tried multiple things but haven't found a solution yet.

I will continue to do so and inform you as soon as I find a solution.

Kind regards,

zepich

**Reply by zepich** [reply, marked as answer]
Hi @davd-edia 

I found a solution for your setup. I tested it with nginx as a reverse proxy, but I think it should also work with IIS.

1. Set the X-Forwarded-Prefix header in the reverse proxy configuration. As mentioned, I've used nginx for the reverse proxy, and my reverse proxy config looks like this:
```config
server {
    listen 443 ssl;

    server_name example.com;

    index index.php;

    location /mosparo/ {
        rewrite /mosparo/(.*) /$1 break;
        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-Prefix /mosparo;
        proxy_pass http://127.0.0.1:8080/;
    }

    ssl_certificate ...;
    ssl_certificate_key ...;
}
```
_Important are the two lines `rewrite /mosparo/(.*) /$1 break;` and `proxy_set_header X-Forwarded-Prefix /mosparo;`. You have to do the same stuff in IIS._

2. In the Docker container, adjust the file `/mosparo/config/packages/framework.yaml` and change the array of trusted headers:
```yaml
...
    trusted_proxies: '%env(resolve:TRUSTED_PROXIES)%'
    trusted_headers: [ 'x-forwarded-for', 'x-forwarded-proto', 'x-forwarded-prefix' ]
...
```
_Add `'x-forwarded-prefix'` to the list of trusted headers._

3. Adjust the font URLs by executing the following command:
```bash
sed -i 's/\/build\///g' /mosparo/public/build/mosparo.css
```
4. Adjust the assets URL by executing the following command:
```bash
sed -i 's/\/build/build/g' /mosparo/public/build/entrypoints.json
```

After that, mosparo should work as expected.

The last two steps are only required since we do not set the `ManifestKeyPrefix` in the Webpack configuration (https://github.com/mosparo/mosparo/blob/master/webpack.config.js#L15). If you build mosparo from the source, you could enable this line, and you do not have to execute steps 3 and 4.

I will try to include steps 2 to 4 in the next minor mosparo version directly, so you don't have to do anything in the future except for the IIS setup (of course).

But with this setup, you do not have to adjust the nginx configuration in the mosparo Docker container.

Please let me know if it worked.

Kind regards,

zepich

**Reply by davd-edia** [reply]
Hey @zepich , that's impressive!! I will test asap. Thanks.

**Reply by zepich** [reply]
Hi @davd-edia 

With the release of version 1.2 today, we've introduced some additional adjustments for this use case.

Please let me know if it's working for you or if you need any additional help.

Kind regards,

zepich
